runtime/src/main/java/org/corfudb/runtime/LogReplicationClient.java

```diff
@@ -0,0 +1,234 @@
+package org.corfudb.runtime;
+
+import com.google.common.base.Strings;
+import com.google.protobuf.Message;
+import org.corfudb.runtime.CorfuCompactorManagement.StringKey;
+import org.corfudb.runtime.LogReplication.ClientInfo;
+import org.corfudb.runtime.LogReplication.ClientInfoKey;
+import org.corfudb.runtime.LogReplication.SinksInfoVal;
+import org.corfudb.runtime.collections.CorfuStore;
+import org.corfudb.runtime.collections.Table;
+import org.corfudb.runtime.collections.TableOptions;
+import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.util.retry.ExponentialBackoffRetry;
+import org.corfudb.util.retry.IRetry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class LogReplicationClient {
+
+    public static final String CORFU_SYSTEM_NAMESPACE = "CorfuSystem";
+    public static final String REGISTRATION_TABLE_NAME = "LogReplicationRegistrationTable";
+    public static final String LR_SOURCE_METADATA_TABLE_NAME = "LogReplicationSourceMetadataTable";
+
+    private static final int SYNC_THRESHOLD = 120;
+
+    private final CorfuStore corfuStore;
+
+    private final String clientName;
+    private final String modelEnum;
+
+    private final Logger log;
+
+    public LogReplicationClient(CorfuRuntime runtime, String clientName) throws InvocationTargetException, NoSuchMethodException, IllegalAccessException {
+        this.corfuStore = new CorfuStore(runtime);
+        this.clientName = clientName;
+
+        // TODO: auto assign model_enum based on the client name
```

Comment:
I'm not sure about this, the client name is not sufficient to know the model. I guess we have to decide if the same LogReplicationClient will be used for all models or we'll have a specific client per model in which case the model is implicit. 


It would be good to know others point of view, but I thought we can have a LogReplicationClient (not specific for a client_name) then have a registerLogicalGroupingClient(), registerRoutingQueueClient() and these registration methods return a specific client type which exposes only the valid calls for each client.

Reply:
IMO having model specific clients gives us some benefits if the behaviors/methods for the two clients are different: we will be able to implement and extend model/client specific methods without keeping in mind to support 2 models. The clients will also have exclusive access to the public methods, instead of exposing all the public methods to all the clients. (e.g. let ROUTING_QUEUE clients have the capability of adding/removing destinations for LOGICAL_GROUP model could be error-prone)